File: GameTemplatePython/Source/Framework/InputComponent.py
import Framework.GameComponent
import pygame

# This component is used by the GameObject to detect if any keys have been pressed or released
class InputComponent(Framework.GameComponent.GameComponent): # It is a child of the GameComponent class
    
    # This is the constructor for the InputComponent. It is called as soon as the InputComponent is created
    def __init__(self, Owner):
        super().__init__(Owner) # This calls the constructor of the base class: 'GameComponent'
        self.Keys = pygame.key.get_pressed() # This stores the current state (pressed or not pressed) of every key on your Keyboard
           
    # These functions are used to query if a Key is currently pressed or released this frame
    def IsKeyPressed(self, KeyCode):
        return self.Keys[KeyCode] == True
    def IsKeyReleased(self, KeyCode):
        return self.Keys[KeyCode] == False
        
    # This stores the current state (pressed or not pressed) of every key on your Keyboard            
    def OnUpdate(self, DeltaTime):
        self.Keys = pygame.key.get_pressed()

File: GameTemplate/Source/Framework/InputComponent.cpp
#include "InputComponent.h"
#include "GameObject.h"
#include "GameFramework.h"

#define KEY_SEEN     1
#define KEY_PRESSED  2

InputComponent::InputComponent(class GameObject* Owner)
    : GameComponent(Owner)
{
    GameFramework::RegisterInputComponent(this);
}


bool InputComponent::IsKeyPressed(int KeyCode) const
{
    int result = Keys[KeyCode] & KEY_PRESSED;
    return result != 0;
}

bool InputComponent::IsKeyReleased(int KeyCode) const
{
    int result = Keys[KeyCode] & KEY_PRESSED;
    return result == 0;
}

bool InputComponent::IsKeyJustPressed(int KeyCode) const
{
    return IsKeyPressed(KeyCode) && Keys[KeyCode] & KEY_SEEN;
}

bool InputComponent::IsKeyJustReleased(int KeyCode) const
{
    return IsKeyReleased(KeyCode) && Keys[KeyCode] & KEY_SEEN;
}

void InputComponent::OnKeyPressed(int KeyCode)
{
    Keys[KeyCode] |= KEY_PRESSED;
    Keys[KeyCode] |= KEY_SEEN;
}

void InputComponent::OnKeyReleased(int KeyCode)
{
    Keys[KeyCode] &= ~KEY_PRESSED;
    Keys[KeyCode] |= KEY_SEEN;
}

void InputComponent::OnUpdate(float DeltaTime)
{
    for (int i = 0; i < ALLEGRO_KEY_MAX; i++)
    {
        Keys[i] &= ~KEY_SEEN;
    }        
}